sympy/functions/elementary/piecewise.py

```diff
@@ -189,7 +190,8 @@ def eval(cls, *_args):
                     _c = c
                     x = free.pop()
                     try:
-                        c = c.as_set().as_relational(x)
+                        if not c.has(Contains):
```

Comment:
Alternatively, you could catch the TypeError or make `Contains.as_relational` return a Relational or raise a NotImplementedError.

Reply:
I think that `as_set` and `as_relational` can be made to work here.

We should have `Contains(expr, S.Integers).as_set() -> S.Integers` if expr is a plain symbol. For other expressions something involving imageset could be used. Maybe ImageSet should always be returned and allowed to evaluate itself.

Reply:
`ImageSet` has no attribute, `as_relational`, so I think it won't work here. Other method can be is to return `self` for both `as_set` and `as_relational` but that's just a workaround. I will use @smichr 's approach.